README.md

```diff
@@ -1,3 +1,226 @@
-# HTTP Middleware
+HTTP Middleware
+===============
 
-PSR-15 interfaces for HTTP middleware.
+1. Summary
+----------
+
+The purpose of this PSR is to provide an interface that defines the formal
+method signature for HTTP Middleware that is compatible with HTTP Messages,
+as defined in PSR-7.
+
+2. Why Bother?
+--------------
+
+The general concept of reusable middleware was popularized by [StackPHP][stackphp].
+Since the release of the HTTP Messages standard, a number of frameworks have
+adopted middleware that uses HTTP Message interfaces.
+
+Agreeing on a formal middleware interface eliminates several problems and
+provides a number of benefits:
+
+* Provides a formal standard for middleware developers to commit to.
+* Eliminates duplication of similar interfaces defined by various frameworks.
+* Avoids minor discrepancies in method signatures.
+* Enables any middleware component to run in any compatible framework.
+
+[stackphp]: http://stackphp.com/
+[express]: http://expressjs.com/en/guide/writing-middleware.html
+
+3. Scope
+--------
+
+## 3.1 Goals
+
+* Create a middleware interface that uses HTTP Messages.
+* Provide a suggested interface for middleware stack containers.
+* Ensure that middleware will not be tied to a specific implementation of HTTP Messages.
+* Implement a middleware signature that is based on best practices.
+
+## 3.2 Non-Goals
+
+* Attempting to define the mechanism by which HTTP responses are created.
+
+4. Approaches
+-------------
+
+There are currently two common approaches to middleware that use HTTP Messages.
+
+### 4.1 Double Pass
+
+The signature used by most middleware implementations has been mostly the same
+and is based on [Express middleware][express], which is defined as:
+
+```
+fn(request, response, next): response
+```
+
+Based on the middleware implementations already used by frameworks that have
+adopted this signature, the following commonalities are observed:
+
+* The middleware is defined as a [callable][php-callable] using `__invoke`.
+* The middleware is passed 3 arguments during invocation:
+  1. A `RequestInterface` implementation.
```

Comment:
If we're talking server-side middleware, this would be a `ServerRequestInterface`. Not sure if it requires updating or not, but would likely lead to less confusion later.